### PR TITLE
Fix issue with unlink teardown behavior

### LINKED ISSUE #762: Service crash occurs after sleep/resume
#### Description: 
A service crash occurs after sleep/resume. This did not occur on Develop prior to autowiring update. Stack trace and branch commit hash attached below.
#### Frequency: 
10/10
#### Found In: 
UPD-Autowiring branch: a0c90e87cc9e174272aa3f1150d90810f8c9e97a
#### Notes: 
Stack Trace:
>	msvcr120.dll!abort() Line 88	C
 	msvcr120.dll!_purecall() Line 59	C
 	LeapSvc.exe!autowiring::signal<void >::operator()<enum VigilanceReason>(enum VigilanceReason &&)	C++
 	LeapSvc.exe!operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,enum LEDMode)	C++
 	LeapSvc.exe!DispatchQueue::DispatchAllEvents(void)	C++
 	LeapSvc.exe!CoreJob::DispatchAllAndClearCurrent(void)	C++
 	LeapSvc.exe!std::_Func_impl<struct std::_Callable_obj<class std::_Bind<0,void,class <lambda_21250f153e69945bed5dd5697e2e3945> >,0>,class std::allocator<class std::_Func_class<void> >,void>::_Do_call(void)	C++
 	LeapSvc.exe!std::_Packaged_state<void >::_Call_immediate(void)	C++
 	LeapSvc.exe!std::_Func_impl<struct std::_Callable_obj<class <lambda_2cb696077b3339a1ae8328ae0c6c0dec>,0>,class std::allocator<class std::_Func_class<void> >,void>::_Do_call(void)	C++
 	LeapSvc.exe!std::_Func_impl<struct std::_Callable_obj<class <lambda_76f314ecb5e1c9167539137e0dc2f350>,0>,class std::allocator<class std::_Func_class<unsigned char> >,unsigned char>::_Do_call(void)	C++
 	LeapSvc.exe!Concurrency::task<unsigned char>::_InitialTaskHandle<void,class <lambda_781036f848fffa6eb15265e2650f848e>,struct Concurrency::details::_TypeSelectorNoAsync>::_Init(struct Concurrency::details::_TypeSelectorNoAsync)	C++
 	LeapSvc.exe!Concurrency::details::_PPLTaskHandle<unsigned char,struct Concurrency::task<unsigned char>::_InitialTaskHandle<void,class <lambda_781036f848fffa6eb15265e2650f848e>,struct Concurrency::details::_TypeSelectorNoAsync>,struct Concurrency::details::_TaskProcHandle>::invoke(void)	C++
 	msvcr120.dll!Concurrency::details::_UnrealizedChore::_UnstructuredChoreWrapper(Concurrency::details::_UnrealizedChore * pChore) Line 293	C++
 	msvcr120.dll!Concurrency::details::InternalContextBase::Dispatch(Concurrency::DispatchState * pDispatchState) Line 1716	C++
 	msvcr120.dll!Concurrency::details::FreeThreadProxy::Dispatch() Line 203	C++
 	msvcr120.dll!Concurrency::details::ThreadProxy::ThreadProxyMain(void * lpParameter) Line 174	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
>

#### ENG Requirements:

### PR DESCRIPTION
**Pull request by codemercenary**
This pull request highlights a problem with the way that unlink teardown occurs, specifically during signal unregistration.  Fix this problem.

- [ ] Fixes #762

### CODE REVIEW

**Comment by codemercenary** [comment]
Any questions on this one @veronicaz41 ?

**Comment by veronicaz41** [comment]
Still looking at this. Will ask you on slack if I have question.

**Comment by veronicaz41** [comment]
Ah, in this case `AutoSignalTest.OneRemovesAll` fails. But why we want `clearall` run  twice? Would it run once be a preferred behavior?
Am I missing something?

**Comment by codemercenary** [comment]
That's the image guarantee.  If I have signal B, and I assert B, then all handlers on B at the time of the call must be asserted, even if one of the handlers removes another handler.

The reason for this is that I don't want nondeterministic cases where the order where signal handlers are called affects how many signal handlers are called.

**Comment by veronicaz41** [comment]
Make sense :+1:  Ready for review.

**Comment by codemercenary** [comment]
This solution is correct, excellent work.  Can you find a way to implement it that does not require any modification to `signal.h`?

**Comment by veronicaz41** [comment]
ok

File: autowiring/auto_signal.h [hunk 1]
```diff
@@ -507,12 +507,14 @@ namespace autowiring {
           state = SignalState::Free;
         } while (!m_state.compare_exchange_weak(state, SignalState::Asserting));
 
-        // Great.  If we got here it's because everyone got away and we're left holding the bag.
-        // We thought we were going to need to pend a callable signal but instead we're still going
-        // to need to invoke it ourselves.
         break;
       }
 
+      if (m_state == SignalState::Free) return;
```
Comment (codemercenary) [inline autowiring/auto_signal.h]: I'm not so sure this is the right answer anymore.  Comparing to `m_state` directly like this is pretty unsafe unless you're doing it in a loop, which we are not.

**Pull request merged**